packages/eslint-plugin/src/rules/no-misused-spread.ts

```diff
@@ -0,0 +1,201 @@
+    const checker = services.program.getTypeChecker();
+
+    function checkArraySpread(node: TSESTree.SpreadElement): void {
+      const type = getConstrainedTypeAtLocation(services, node.argument);
+
+      if (isString(type)) {
+        context.report({
+          node,
+          messageId: 'noStringSpreadInArray',
+        });
+
+        return;
+      }
+    }
+
+    function checkObjectSpread(node: TSESTree.SpreadElement): void {
+      const type = getConstrainedTypeAtLocation(services, node.argument);
+
+      if (isArray(type, checker)) {
+        context.report({
+          node,
+          messageId: 'noSpreadInObject',
+          data: {
+            type: 'Array',
+          },
+        });
+
+        return;
+      }
+
+      if (isPromise(services.program, type)) {
+        context.report({
+          node,
+          messageId: 'noSpreadInObject',
+          data: {
+            type: 'Promise',
+          },
+        });
+
+        return;
+      }
+
+      if (isFunctionWithoutProps(type)) {
+        context.report({
+          node,
+          messageId: 'noFunctionSpreadInObject',
+        });
+
+        return;
+      }
+
+      if (isIterable(type, checker) && !isString(type)) {
+        context.report({
+          node,
+          messageId: 'noSpreadInObject',
+          data: {
+            type: 'Iterable',
+          },
+        });
+
+        return;
+      }
+
+      if (!options.allowClassInstances && isClassInstance(type)) {
+        context.report({
+          node,
+          messageId: 'noClassInstanceSpreadInObject',
+        });
+
+        return;
+      }
+
+      if (node.argument.type === AST_NODE_TYPES.ClassExpression) {
+        context.report({
+          node,
+          messageId: 'noClassDeclarationSpreadInObject',
+        });
+
+        return;
+      }
+    }
+
+    return {
+      'ArrayExpression > SpreadElement': checkArraySpread,
+      'ObjectExpression > SpreadElement': checkObjectSpread,
+    };
+  },
+});
+
+function isArray(type: ts.Type, checker: ts.TypeChecker): boolean {
+  if (type.isUnion() || type.isIntersection()) {
+    return type.types.some(t => isArray(t, checker));
+  }
+
+  return checker.isArrayType(type) || checker.isTupleType(type);
+}
+
+function isIterable(type: ts.Type, checker: ts.TypeChecker): boolean {
+  return tsutils
+    .typeParts(type)
+    .some(
+      t => !!tsutils.getWellKnownSymbolPropertyOfType(t, 'iterator', checker),
+    );
+}
+
+function isString(type: ts.Type): boolean {
+  return tsutils
+    .typeParts(type)
+    .some(t => !!isTypeFlagSet(t, ts.TypeFlags.StringLike));
+}
+
+function isFunctionWithoutProps(type: ts.Type): boolean {
+  return tsutils
+    .typeParts(type)
+    .some(
+      t => t.getCallSignatures().length > 0 && t.getProperties().length === 0,
+    );
+}
+
+function isPromise(program: ts.Program, type: ts.Type): boolean {
+  return tsutils.typeParts(type).some(t => isPromiseLike(program, t));
+}
```

Comment:
```suggestion
```

`isPromiseLike` already contains the type part iteration logic :smile: I think there is no need to do `typeParts` here!

https://github.com/typescript-eslint/typescript-eslint/blob/f248e689b04bdfd81429977a731ba6b8ad82622e/packages/type-utils/src/builtinSymbolLikes.ts#L132-L141

Reply:
seems like it doesn't catch this case:

```TS
declare const maybePromise: Promise<number> | { a: number };
const o = { ...maybePromise };
```

and I think we should catch it, because it means we might spread a promise in runtime, right?

anyway, see this comment:
https://github.com/typescript-eslint/typescript-eslint/pull/8509#discussion_r1528451293

I'll be glad to get some help here with removing this check completely 😄 
Wasn't able to figure out why `isClassOrInterface` didn't work here